google/cloud/internal/url_encode.cc

```diff
@@ -24,35 +23,54 @@ namespace internal {
 
 namespace {
 
-using UrlMappings =
-    std::array<std::pair<absl::string_view, absl::string_view>, 25>;
-
-constexpr auto kEncodeMappings = UrlMappings{{
-    {" ", "%20"}, {"\"", "%22"}, {"#", "%23"},  {"$", "%24"}, {"%", "%25"},
-    {"&", "%26"}, {"+", "%2B"},  {",", "%2C"},  {"/", "%2F"}, {":", "%3A"},
-    {";", "%3B"}, {"<", "%3C"},  {"=", "%3D"},  {">", "%3E"}, {"?", "%3F"},
-    {"@", "%40"}, {"[", "%5B"},  {"\\", "%5C"}, {"]", "%5D"}, {"^", "%5E"},
-    {"`", "%60"}, {"{", "%7B"},  {"|", "%7C"},  {"}", "%7D"}, {"\177", "%7F"},
-}};
-
-// Until we require C++17, it isn't worth trying to eliminate the duplication
-// using a constexpr function to produce kDecodeMappings from kEncodeMappings.
-constexpr auto kDecodeMappings = UrlMappings{{
-    {"%20", " "}, {"%22", "\""}, {"%23", "#"},  {"%24", "$"}, {"%25", "%"},
-    {"%26", "&"}, {"%2B", "+"},  {"%2C", ","},  {"%2F", "/"}, {"%3A", ":"},
-    {"%3B", ";"}, {"%3C", "<"},  {"%3D", "="},  {"%3E", ">"}, {"%3F", "?"},
-    {"%40", "@"}, {"%5B", "["},  {"%5C", "\\"}, {"%5D", "]"}, {"%5E", "^"},
-    {"%60", "`"}, {"%7B", "{"},  {"%7C", "|"},  {"%7D", "}"}, {"%7F", "\177"},
-}};
+bool WeNeedToEscape(unsigned char c) {
+  static auto const* const kSet = new std::set<char>{
+      ' ', '\"', '#', '$', '%', '&',  '+', ',', '/', ':', ';', '<',
+      '=', '>',  '?', '@', '[', '\\', ']', '^', '`', '{', '|', '}',
+  };
+  return std::isprint(c) == 0 || kSet->find(c) != kSet->end();
+}
+
+// Returns 0-15 if c is in [0-9A-Fa-f], and -1 otherwise.
+int ParseDigitHex(char c) {
+  if ('0' <= c && c <= '9') return c - '0';
+  if ('A' <= c && c <= 'F') return c - 'A' + 0xA;
+  if ('a' <= c && c <= 'f') return c - 'a' + 0xa;
+  return -1;
+}
 
 }  // namespace
 
 std::string UrlEncode(absl::string_view value) {
-  return absl::StrReplaceAll(value, kEncodeMappings);
+  std::string s;
+  auto constexpr kDigits = "0123456789ABCDEF";
+  for (unsigned char c : value) {
+    if (WeNeedToEscape(c)) {
+      s.push_back('%');
```

Comment:
You should choose between `s.push_back(c)` and `s += c` (used below).

Reply:
True, fixed.